File: codes/2024/25-Excellence-BNU-99-21-43 (C++)/workspace/source/backend/operand/ExStackVarContent.cpp
// ExStackVarContent.cpp

#include "ExStackVarContent.h"

namespace Backend {

ExStackVarContent::ExStackVarContent(AsmOperandMemoryAddress* address, int size, bool isS0Based)
    : AsmOperandStackVariable(address, size, isS0Based) { setType(OperandType::EX_STACK_VAR_CONTENT); }

ExStackVarContent* ExStackVarContent::transform(AsmOperandStackVariable* stackVar) {
    return new ExStackVarContent(stackVar->getAddress(), stackVar->getSize(), stackVar->isS0Based());
}

ExStackVarContent* ExStackVarContent::transform(AsmOperandStackVariable* stackVar, AsmOperandMemoryAddress *address) {
    return new ExStackVarContent(address, stackVar->getSize(), stackVar->isS0Based());
}

AsmOperandBasic* ExStackVarContent::add(int diff) {
    if (isS0Based() && getAddress()->getOffset() < 0) {
        return new ExStackVarContent(getAddress()->addOffset(diff), getSize(), true);
    } else {
        return this;
    }
}

AsmOperandBasic* ExStackVarContent::withRegister(AsmOperandRegister* reg) const {
    return new ExStackVarContent(static_cast<AsmOperandMemoryAddress*>(getAddress()->withRegister(reg)), getSize(), isS0Based());
}

} // namespace Backend

File: codes/2023/2-First-BIT (Java)/src/main/java/cn/edu/bit/newnewcc/backend/asm/operand/ExStackVarContent.java
package cn.edu.bit.newnewcc.backend.asm.operand;

/**
 * 栈变量内容，专用于栈帧重新分配过程.
 */
public class ExStackVarContent extends StackVar implements ExStackVarAdd {
    private ExStackVarContent(MemoryAddress address, int size, boolean isS0Based) {
        super(address, size, isS0Based);
    }
    public static ExStackVarContent transform(StackVar stackVar) {
        return new ExStackVarContent(stackVar.getAddress(), stackVar.getSize(), stackVar.isS0Based());
    }
    public static ExStackVarContent transform(StackVar stackVar, MemoryAddress address) {
        return new ExStackVarContent(address, stackVar.getSize(), stackVar.isS0Based());
    }

    @Override
    public AsmOperand add(int diff) {
        if (isS0Based() && address.getOffset() < 0) {
            return new ExStackVarContent(getAddress().addOffset(diff), getSize(), true);
        } else {
            return this;
        }
    }

    @Override
    public ExStackVarContent withRegister(Register register) {
        return new ExStackVarContent(address.withRegister(register), size, isS0);
    }
}
